functionary/builder/utils.py

```diff
@@ -105,6 +255,109 @@ def initiate_build(
     return build
 
 
+def create_package_from_definition(
+    package_definition: dict, environment: "Environment"
+) -> Package:
+    """Create a package from a package definition"""
+    # TODO: Manually parsing for now, but this should be codified somewhere, with
+    #       the parsing informed by package_definition_version.
+    package_obj = Package(
+        environment=environment,
+        name=package_definition.get("name"),
+    )
+
+    package_obj.display_name = package_definition.get("display_name")
+    package_obj.summary = package_definition.get("summary")
+    package_obj.description = package_definition.get("description")
+    package_obj.language = package_definition.get("language")
+
+    package_obj.save()
+
+    return package_obj
+
+
+@app.task
+def build_package(build_id: "UUID") -> None:
+    """Retrieve the resources for Build and use them to build and push the package
+    docker image. Also creates BuildLog with build/push information.
+
+    Args:
+        build_id: ID of the build being executed
+
+    Returns:
+        None
+    """
+    build = Build.objects.select_related("environment").get(id=build_id)
+    build_resource = BuildResource.objects.get(build=build)
+
+    logger.info(f"Starting build {build.id}")
+
+    try:
+        build, log = _build_package(build, build_resource)
+        build.status = Build.COMPLETE
+        _create_build_log(build, log)
+        _update_build_status(build)
+        logger.info(f"Build {build_id} COMPLETE")
+    except BuilderException as err:
+        logger.error(f"Encountered build error: {err}")
+        build.status = Build.ERROR
+        _create_build_log(build, err)
+        _update_build_status(build)
+    except DockerSocketConnectionError as err:
```

Comment:
Is there a reason why you wouldn't just record this as a build error like with the other exceptions?  It would still be useful to see that the build failed and why.

Also, if you do that, I think you can restructure this to just have a finally at the end that updates the build:

```python
    try:
        build, log = _build_package(build, build_resource)
        build.status = Build.COMPLETE
        logger.info(f"Build {build_id} COMPLETE")
    except Exception as err:
        log = str(err)
        build.status = Build.ERROR
        logger.error(f"Build {build_id} failed: {err}")
    finally:
        _create_build_log(build, log)
        _update_build_status(build)
```

I also wonder if having `_build_package` return the log is the correct thing to do, versus passing it an object that it can build up as the log.  Even just a list that it can keep adding entries to.  The reason I'm thinking that is because as is, if we don't make it to the end of `_build_package`, any of the potentially useful log entries up to that point are lost because we didn't reach the return.  If we instead were appending entries to a list inside `_build_package`, that list would be intact even if we never reach the return.  Then you could append the exception output to the rest of the log and we'd have the full picture of how far the build got before failing.

Reply:
Yea, that makes sense. That's originally what I was trying to accomplish with my Go-style iteration